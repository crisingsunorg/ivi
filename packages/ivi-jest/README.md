`ivi-jest` provides a collection of useful tools for testing with [Jest](https://jestjs.io/).

## Motivation

Jest resets global state for every test module, but all tests in a test module share the same global state. This package
provides a set of tools to make it easier to write isolated unit tests.

## API

### DOM

`useResetDOMGlobalEventListeners()` tracks event handlers for a window and document objects and removes them after
each test.

`useResetDOM()` enables following resets:

- `useResetDOMGlobalEventListeners()`

`useDOMElement()` creates a DOM element and optionally mounts it to the document body.

`useRequestAnimationFrame()` mocks `requestAnimationFrame()` and `cancelAnimationFrame()`.

### Modules

`useResetModules()` resets modules before each test.

`useModule()` imports module before each test.

### Mock

`useMockFn()` creates a mock function and resets it after each test.

`useSpyOn()` creates a spy object and resets it after each test.

### Environment

`useEnv()` overrides `process.env` before each tests and restores to the original value after each test.

### Promise

`usePromiseQueue()` adds support for waiting until promise tasks are executed.

### Timers

`useTimers()` mocks timers.

### Computed Value

`useComputedValue()` creates a value that will be recomputed before each test.
