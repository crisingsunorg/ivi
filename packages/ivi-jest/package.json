{
  "private": true,
  "name": "ivi-jest",
  "version": "0.0.1",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "sideEffects": false,
  "files": [
    "dist",
    "src",
    "!dist/**/*.tsbuildinfo",
    "!src/**/__tests__",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run dist",
    "clean": "rm -rf ./dist",
    "dist": "npm run dist:main && npm run dist:module",
    "dist:main": "../../node_modules/.bin/tsc -p ./tsconfig.build.main.json",
    "dist:module": "../../node_modules/.bin/tsc -p ./tsconfig.build.module.json"
  },
  "dependencies": {
    "jsdom": "^14.0.0"
  },
  "peerDependencies": {},
  "devDependencies": {},
  "description": "Jest utils for ivi library",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ivi"
  ],
  "homepage": "https://github.com/localvoid/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/localvoid/ivi"
  },
  "bugs": {
    "url": "https://github.com/localvoid/ivi/issues"
  },
  "github": "https://github.com/localvoid/ivi"
}
