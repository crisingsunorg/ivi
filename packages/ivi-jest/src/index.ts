export { useComputedValue } from "./value";
export { useEnv } from "./env";
export { useMockFn, useSpyOn } from "./mock";
export { useResetModules, useModule, useIVI, useHTML, useSVG, useTest } from "./module";
export { useResetDOMGlobalEventListeners, useResetDOM, useDOMElement } from "./dom";
