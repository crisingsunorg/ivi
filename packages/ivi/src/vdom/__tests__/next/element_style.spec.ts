import { useResetJSDOM, useResetModules, useSpyOn, useDOMElement, useHTML, useSVG, useTest } from "ivi-jest";

useResetJSDOM();
useResetModules();
const setProperty = useSpyOn(() => CSSStyleDeclaration.prototype, "setProperty");
const c = useDOMElement();
const t = useTest();
const _ = void 0;

describe("element style", () => {
  describe("HTML", () => {
    const h = useHTML();
    const r = (style?: {}) => t.render<HTMLElement>(h.div(_, { style }), c()).domNode!;

    describe("mount", () => {
      test("undefined", () => {
        const n = r();
        expect(setProperty().mock.calls.length).toBe(0);
        expect(n).toMatchSnapshot();
      });

      test("property with undefined value", () => {
        const n = r({ top: void 0 });
        expect(setProperty().mock.calls.length).toBe(0);
        expect(n).toMatchSnapshot();
      });

      test("one property", () => {
        const n = r({ top: "10px" });
        expect(setProperty().mock.calls.length).toBe(1);
        expect(n).toMatchSnapshot();
      });

      test("two properties", () => {
        const n = r({ top: "10px", left: "20px" });
        expect(setProperty().mock.calls.length).toBe(2);
        expect(n).toMatchSnapshot();
      });
    });
  });

  describe("SVG", () => {
    const s = useSVG();
    const r = (style?: {}) => t.render<SVGElement>(s.circle(_, { style }), c()).domNode!;

    describe("mount", () => {
      test("undefined", () => {
        const n = r();
        expect(setProperty().mock.calls.length).toBe(0);
        expect(n).toMatchSnapshot();
      });

      test("property with undefined value", () => {
        const n = r({ top: void 0 });
        expect(setProperty().mock.calls.length).toBe(0);
        expect(n).toMatchSnapshot();
      });

      test("one property", () => {
        const n = r({ top: "10px" });
        expect(setProperty().mock.calls.length).toBe(1);
        expect(n).toMatchSnapshot();
      });

      test("two properties", () => {
        const n = r({ top: "10px", left: "20px" });
        expect(setProperty().mock.calls.length).toBe(2);
        expect(n).toMatchSnapshot();
      });
    });
  });
});
