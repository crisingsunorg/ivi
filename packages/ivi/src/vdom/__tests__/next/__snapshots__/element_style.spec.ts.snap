// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`element style HTML mount empty style object 1`] = `<div />`;

exports[`element style HTML mount one property 1`] = `
<div
  style="top: 10px;"
/>
`;

exports[`element style HTML mount property with undefined value 1`] = `<div />`;

exports[`element style HTML mount two properties 1`] = `
<div
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style HTML mount undefined 1`] = `<div />`;

exports[`element style HTML update empty style object to empty style object 1`] = `<div />`;

exports[`element style HTML update empty style object to one property 1`] = `
<div
  style="top: 10px;"
/>
`;

exports[`element style HTML update empty style object to property with undefined value 1`] = `<div />`;

exports[`element style HTML update empty style object to two properties 1`] = `
<div
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style HTML update empty style object to undefined 1`] = `<div />`;

exports[`element style HTML update one property different property 1`] = `
<div
  style="left: 20px;"
/>
`;

exports[`element style HTML update one property empty style object 1`] = `
<div
  style=""
/>
`;

exports[`element style HTML update one property property with different value 1`] = `
<div
  style="top: 20px;"
/>
`;

exports[`element style HTML update one property property with same value 1`] = `
<div
  style="top: 10px;"
/>
`;

exports[`element style HTML update one property property with undefined value 1`] = `
<div
  style=""
/>
`;

exports[`element style HTML update one property two properties 1`] = `
<div
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style HTML update one property undefined 1`] = `
<div
  style=""
/>
`;

exports[`element style HTML update property with undefined value empty style object 1`] = `<div />`;

exports[`element style HTML update property with undefined value one property 1`] = `
<div
  style="top: 10px;"
/>
`;

exports[`element style HTML update property with undefined value property with undefined value 1`] = `<div />`;

exports[`element style HTML update property with undefined value two properties 1`] = `
<div
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style HTML update property with undefined value undefined 1`] = `<div />`;

exports[`element style HTML update two properties empty style object 1`] = `
<div
  style=""
/>
`;

exports[`element style HTML update two properties one property with different value 1`] = `
<div
  style="left: 30px;"
/>
`;

exports[`element style HTML update two properties one property with same value 1`] = `
<div
  style="left: 20px;"
/>
`;

exports[`element style HTML update two properties properties with different values 1`] = `
<div
  style="top: 30px; left: 40px;"
/>
`;

exports[`element style HTML update two properties properties with same values 1`] = `
<div
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style HTML update two properties property with undefined value 1`] = `
<div
  style=""
/>
`;

exports[`element style HTML update two properties undefined 1`] = `
<div
  style=""
/>
`;

exports[`element style HTML update undefined to empty style object 1`] = `<div />`;

exports[`element style HTML update undefined to one property 1`] = `
<div
  style="top: 10px;"
/>
`;

exports[`element style HTML update undefined to property with undefined value 1`] = `<div />`;

exports[`element style HTML update undefined to two properties 1`] = `
<div
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style HTML update undefined to undefined 1`] = `<div />`;

exports[`element style SVG mount empty style object 1`] = `<circle />`;

exports[`element style SVG mount one property 1`] = `
<circle
  style="top: 10px;"
/>
`;

exports[`element style SVG mount property with undefined value 1`] = `<circle />`;

exports[`element style SVG mount two properties 1`] = `
<circle
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style SVG mount undefined 1`] = `<circle />`;

exports[`element style SVG update empty style object to empty style object 1`] = `<circle />`;

exports[`element style SVG update empty style object to one property 1`] = `
<circle
  style="top: 10px;"
/>
`;

exports[`element style SVG update empty style object to property with undefined value 1`] = `<circle />`;

exports[`element style SVG update empty style object to two properties 1`] = `
<circle
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style SVG update empty style object to undefined 1`] = `<circle />`;

exports[`element style SVG update one property different property 1`] = `
<circle
  style="left: 20px;"
/>
`;

exports[`element style SVG update one property empty style object 1`] = `
<circle
  style=""
/>
`;

exports[`element style SVG update one property property with different value 1`] = `
<circle
  style="top: 20px;"
/>
`;

exports[`element style SVG update one property property with same value 1`] = `
<circle
  style="top: 10px;"
/>
`;

exports[`element style SVG update one property property with undefined value 1`] = `
<circle
  style=""
/>
`;

exports[`element style SVG update one property two properties 1`] = `
<circle
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style SVG update one property undefined 1`] = `
<circle
  style=""
/>
`;

exports[`element style SVG update property with undefined value empty style object 1`] = `<circle />`;

exports[`element style SVG update property with undefined value one property 1`] = `
<circle
  style="top: 10px;"
/>
`;

exports[`element style SVG update property with undefined value property with undefined value 1`] = `<circle />`;

exports[`element style SVG update property with undefined value two properties 1`] = `
<circle
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style SVG update property with undefined value undefined 1`] = `<circle />`;

exports[`element style SVG update two properties empty style object 1`] = `
<circle
  style=""
/>
`;

exports[`element style SVG update two properties one property with different value 1`] = `
<circle
  style="left: 30px;"
/>
`;

exports[`element style SVG update two properties one property with same value 1`] = `
<circle
  style="left: 20px;"
/>
`;

exports[`element style SVG update two properties properties with different values 1`] = `
<circle
  style="top: 30px; left: 40px;"
/>
`;

exports[`element style SVG update two properties properties with same values 1`] = `
<circle
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style SVG update two properties property with undefined value 1`] = `
<circle
  style=""
/>
`;

exports[`element style SVG update two properties undefined 1`] = `
<circle
  style=""
/>
`;

exports[`element style SVG update undefined to empty style object 1`] = `<circle />`;

exports[`element style SVG update undefined to one property 1`] = `
<circle
  style="top: 10px;"
/>
`;

exports[`element style SVG update undefined to property with undefined value 1`] = `<circle />`;

exports[`element style SVG update undefined to two properties 1`] = `
<circle
  style="top: 10px; left: 20px;"
/>
`;

exports[`element style SVG update undefined to undefined 1`] = `<circle />`;
